powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationBehavior.java

```diff
@@ -0,0 +1,469 @@
+
+            } else {
+                throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_APPROVE_FAILURE);
+            }
+        } else {
+            throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_APPROVE_FAILURE);
+        }
+    }
+
+    public OperationUserActionResponse rejectOperation(OperationRejectRequest request) throws GenericServiceException {
+        final Date currentTimestamp = new Date();
+
+        final String operationId = request.getOperationId();
+        final String userId = request.getUserId();
+        final Long applicationId = request.getApplicationId();
+
+        // TODO: We might need a lock here!!!
+        final Optional<OperationEntity> operationOptional = operationRepository.findOperation(operationId);
+        if (operationOptional.isPresent()) {
+            final OperationEntity operationEntity = expireOperation(operationOptional.get(), currentTimestamp);
+            if (OperationStatusDo.PENDING.equals(operationEntity.getStatus())) {
+
+                if (operationEntity.getUserId().equals(userId) // correct user rejects the operation
+                        && operationEntity.getApplicationId().equals(applicationId)) { // operation is rejected by the expected application
+
+                    // Approve the operation
+                    operationEntity.setStatus(OperationStatusDo.REJECTED);
+                    operationEntity.setTimestampFinalized(currentTimestamp);
+
+                    final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                    final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                    OperationUserActionResponse response = new OperationUserActionResponse();
+                    response.setResult(UserActionResult.REJECTED);
+                    response.setOperation(operationDetailResponse);
+                    return response;
+                } else {
+                    final OperationDetailResponse operationDetailResponse = convertFromEntity(operationEntity);
+                    OperationUserActionResponse response = new OperationUserActionResponse();
+                    response.setResult(UserActionResult.REJECT_FAILED);
+                    response.setOperation(operationDetailResponse);
+                    return response;
+                }
+
+            } else {
+                throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_REJECT_FAILURE);
+            }
+        } else {
+            throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_REJECT_FAILURE);
+        }
+    }
+
+    public OperationUserActionResponse failApprovalOperation(OperationFailApprovalRequest request) throws GenericServiceException {
+        final Date currentTimestamp = new Date();
+
+        final String operationId = request.getOperationId();
+
+        // TODO: We might need a lock here!!!
```

Comment:
Indeed, I think that the lock would be useful here because the processing changes the operation state.

Reply:
👍 I used the lock function for all methods that modify an operation state.